src/crimson/osd/osd.cc

```diff
@@ -823,6 +830,104 @@ seastar::future<> OSD::handle_pg_log(ceph::net::Connection* conn,
+	  m->epoch < past_intervals.get_bounds().second) {
+        logger().error("got pg_create on {} epoch {} and unmatched "
+                       "past_intervals {} (history {})",
+                       pgid, m->epoch, past_intervals, pg_history);
+      } if (const auto found = pgs.find(pgid); found != pgs.end()) {
+        logger().info("pg {} already exists. Ignoring creation request", pgid);
+      } else if (osdmap->is_up_acting_osd_shard(pgid, whoami)) {
+        logger().info("pg {} does not exist. Creating", pgid);
+        // XXX: peering machinery bypassed. Handling create_info directly.
+        return handle_pg_create_info(osdmap, std::make_unique<PGCreateInfo>(
+	  pgid, osdmap->get_epoch(), pg_history, past_intervals, true));
+      }
+      return seastar::now();
+  }).then([m, this] {
+    // TODO: pending_creates_from_mon
+    last_pg_create_epoch = m->epoch;
+  });
+}
+
+seastar::future<> OSD::handle_pg_create_info(cached_map_t osdmap,
+                                             std::unique_ptr<PGCreateInfo> info)
+{ 
+  if (info->by_mon) {
+    const auto* pool = osdmap->get_pg_pool(info->pgid.pgid.pool());
+    if (!pool) {
+      logger().info("{} ignoring {}, pool dne", __func__, info->pgid);
+      return seastar::now();
+    }
+    if (osdmap->require_osd_release >= CEPH_RELEASE_NAUTILUS &&
+	!pool->has_flag(pg_pool_t::FLAG_CREATING)) {
+      // this ensures we do not process old creating messages after the
+      // pool's initial pgs have been created (and pg are subsequently
+      // allowed to split or merge).
+      logger().debug("{} dropping {} create, pool does not have CREATING flag",
+                     __func__, info->pgid);
+      return seastar::now();
+    }
+  }
+
+  using ec_profile_t = map<string,string>;
+  return get_map(info->epoch).then(
+    [osdmap,info=std::move(info),this](auto&& create_map) {
+      if (!create_map->have_pg_pool(info->pgid.pool())) {
+        // pool was deleted; grab final pg_pool_t off disk.
+        return seastar::now();
+      }
+
+      pg_pool_t pi = *create_map->get_pg_pool(info->pgid.pool());
+
+      const coll_t coll{info->pgid};
+      ceph::os::Transaction txn{};
+      txn.create_collection(coll, info->pgid.get_split_bits(pi.get_pg_num()));
+
+      auto ec_profile = !pi.is_erasure() ? ec_profile_t{} :
+        create_map->get_erasure_code_profile(pi.erasure_code_profile);
+      auto backend = PGBackend::create(info->pgid, pi, store.get(), ec_profile);
+      Ref<PG> pg{new PG{info->pgid,
+                        pg_shard_t{whoami, info->pgid.shard},
+                        std::move(pi),
+                        create_map->get_pool_name(info->pgid.pool()),
+                        std::move(backend),
+                        osdmap,
+                        cluster_msgr}};
+      pg->initialize(std::move(info->past_intervals));
+      pg->handle_activate_map();
+      pgs.emplace(info->pgid, pg);
+
+      if (!txn.empty()) {
```

Comment:
under what circumstances `txn` could be empty? we always call `txn.create_collection(...)` right after `txn` is defined.